stlab/concurrency/router.hpp

```diff
@@ -0,0 +1,137 @@
+/*
+    Copyright 2015 Adobe
+    Distributed under the Boost Software License, Version 1.0.
+    (See accompanying file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+*/
+
+/**************************************************************************************************/
+
+#ifndef STLAB_CONCURRENCY_ROUTER_HPP
+#define STLAB_CONCURRENCY_ROUTER_HPP
+
+#include <algorithm>
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include <stlab/concurrency/channel.hpp>
+
+/**************************************************************************************************/
+
+namespace stlab {
+
+/**************************************************************************************************/
+
+inline namespace v1 {
+
+/**************************************************************************************************/
+
+template <class T, class K>
+class router {
+    using channel_pair = std::pair<stlab::sender<T>, stlab::receiver<T>>;
+    using route_pair = std::pair<K, channel_pair>;
+    using routes = std::vector<route_pair>;
+
+public:
+    template <class E, class F>
+    router(E executor, F router_func) : _self{std::make_shared<model<E, F>>(std::move(executor), std::move(router_func))} {}
+
+    void set_ready() { _self->set_ready(); }
+    stlab::receiver<T> get_route(K key) { return _self->get_route(std::move(key)); }
```

Comment:
I don't like that the function is not const. The function does more than one thing: It returns an existing route or if it does not exist, then it creates it and returns it then. Do you see my point? 

Is the namespace stlab here and at other places necessary? We are already within stlab.